### DISCUSSION TITLE
try to download a file from deta drive through a endpoint

### DISCUSSION THREAD

**Discussion by ooralas** [deta/cloud-docs#362] [Help] [answered]
I have a nodejs-micro. i have already uploaded a couple of PDF files
, which worked without any problems. When I try to
make the files available for download through an endpoint, 
it did not work.
I get a BLOB object from the drive. If I do this locally (see the code), I can use this code
```javascript 
const reports = deta.Drive('reports')
const file = await reports.get(fileName).then(a => a.arrayBuffer())
  fs.writeFileSync('./' + fileName, Buffer.from(file))
  res.download(fileName)
```
where i convert the BLOB object to a PDF file, and then make it available for Download. But when i deployed my micro, this didn't work because I'm probably not allowed to access the filesystem on deta.
So my question is, as described above, can I make a file, that is in Drive, available for download through an endpoint ?

**Comment by jnsougata** [comment]
Deta drive doesn't support that. But yes you can store the file in micro at `/tmp` dir as that's only runtime writable

**Comment by ooralas** [comment]
@jnsougata thanks.
which `fs` Methods can i use in `/tmp`? i am getting "Bad Gateway" when i try to `fs.unlinkSync()`, in addition when i tried to `res.download()` a pdf file, which i stored on `/tmp` locally and deployed on my micro, i got an empty pdf file with same page numbers like the original file.
what should be the problem ?

**Comment by abdelhai** [comment]
@ooralas micros do not persist data. you can't download things there and expect them to exist the next time.
`/tmp` are meant to be temporary place to hold files while you process them during a specific request.

to make it a bit more clear, if you download a file during with one request and tried to read it in the other, it's almost always not there.


to be able to help you, please provide more details on what you're trying to achieve on a high level.

**Reply by ooralas** [reply]
thanks for you answer.

I have 23 pdf files that I want to make available for download through a specific URL. I have saved the files on Drive. 
By the parameter, I know which file to download, which I then retrieve from Drive. 
From Drive I get a buffer, which I handle as follows, and then send the file for download. Locally it works, I get the file and I can open it. But when I deploy my micro, I get the file with the correct file size and the correct page count, but the pages are blank.

```javascript
route.get('/p/:name', async (req, res) => {
  const blob = await berichte.get('wb1.pdf')
  const buffer = await blob.arrayBuffer()
  console.log(buffer)
  res.setHeader('Content-Type', 'application/pdf')
  res.setHeader('Content-Disposition', 'attachment; filename=wb1.pdf')
  res.send(Buffer.from(buffer))
}
 ```

p.s.: the files remain in /tmp. 
i saved 3 files there yesterday and they are there since then.

**Reply by ooralas** [reply]
@abdelhai

**Reply by aavshr** [reply]
hi @ooralas the code looks good to me, can you please try to put the env var `BINARY_CONTENT_TYPES=application/pdf` for the Micro and try again? The issue might be because of this, more here: https://docs.deta.sh/docs/common_issues#nodejs-micros-cannot-serve-binary-files

Also, please bear in mind that Micros have a payload size limit of 5.5 Mb so larger files can't be served from the Micro.

**Reply by ooralas** [reply]
it worked thank you. 

Can you please tell me how can i clean `/tmp` directory on my micro ? i tried `fs.unlinkSync('./tmp/wb2.pdf')` i didn't work. i got a bad gateway error, which `fs` functions can i use on micro ?

**Reply by aavshr** [reply]
`/tmp` is automatically cleaned @ooralas . 

I'm sorry I'm not fully aware what kind of APIs `fs` has but all functions that allow you to `read` from a file should work, and functions that need to `write` to a file only work for files in `/tmp`.  Also you can try to look at the logs when there is a `Bad Gateway` error to try to see what exactly is the issue.

**Comment by ooralas** [comment, marked as answer]
@aavshr thanks for your answer, after i update the `.env` file it worked.
> hi @ooralas the code looks good to me, can you please try to put the env var `BINARY_CONTENT_TYPES=application/pdf` for the Micro and try again? The issue might be because of this, more here: https://docs.deta.sh/docs/common_issues#nodejs-micros-cannot-serve-binary-files
> 
> Also, please bear in mind that Micros have a payload size limit of 5.5 Mb so larger files can't be served from the Micro.

@aavshr i couldn't mark you answer in the thread as an answer, so i will do it hier :).
